src/python_testing/TC_OCC_2_1.py

```diff
@@ -0,0 +1,253 @@
+            TestStep(11, "Read UltrasonicUnoccupiedToOccupiedDelay attribute, if supported"),
+            TestStep(12, "Read UltrasonicUnoccupiedToOccupiedThreshold attribute, if supported"),
+            TestStep(13, "Read PhysicalContactOccupiedToUnoccupiedDelay attribute, if supported"),
+            TestStep(14, "Read PhysicalContactUnoccupiedToOccupiedDelay attribute, if supported"),
+            TestStep(15, "Read PhysicalContactUnoccupiedToOccupiedThreshold attribute, if supported")
+        ]
+        return steps
+
+    def pics_TC_OCC_2_1(self) -> list[str]:
+        pics = [
+            "OCC.S",
+        ]
+        return pics
+
+    @async_test_body
+    async def test_TC_OCC_2_1(self):
+
+        endpoint = self.user_params.get("endpoint", 1)
+
+        self.step(1)
+        attributes = Clusters.OccupancySensing.Attributes
+        attribute_list = await self.read_occ_attribute_expect_success(endpoint=endpoint, attribute=attributes.AttributeList)
+
+        self.step(2)
+        if attributes.Occupancy.attribute_id in attribute_list:
+            occupancy_dut = await self.read_occ_attribute_expect_success(endpoint=endpoint, attribute=attributes.Occupancy)
+            asserts.assert_less_equal(occupancy_dut, 0b00000001, "Occupancy attribute is not in valid range")
+        else:
+            logging.info("Occupancy attribute is a mandatory attribute. Test step fails.")
+            asserts.fail("Missing mandatory attribute Occupancy")
+
+        self.step(3)
+        if attributes.OccupancySensorType.attribute_id in attribute_list:
+            occupancy_sensor_type_dut = await self.read_occ_attribute_expect_success(endpoint=endpoint, attribute=attributes.OccupancySensorType)
+            asserts.assert_less(occupancy_sensor_type_dut, Clusters.Objects.OccupancySensing.Enums.OccupancySensorTypeEnum.kUnknownEnumValue,
+                                "OccupancySensorType is not in valid range")
+        else:
+            logging.info("OccupancySensorType attribute is a mandatory attribute. Test step fails")
+            asserts.fail("Missing mandatory attribute OccupancySensorType")
+
+        self.step(4)
+        if attributes.OccupancySensorTypeBitmap.attribute_id in attribute_list:
+            occupancy_sensor_type_bitmap_dut = await self.read_occ_attribute_expect_success(endpoint=endpoint, attribute=attributes.OccupancySensorTypeBitmap)
+            asserts.assert_less_equal(occupancy_sensor_type_bitmap_dut, 0b00000111, "OccupancySensorTypeBitmap attribute is not in valid range")
+        else:
+            logging.info("OccupancySensorTypeBitmap attribute is a mandatory attribute. Test step fails")
+            asserts.fail("Missing mandatory attribute OccupancySensorTypeBitmap")
+
+        self.step(5)
+        if attributes.HoldTimeLimits.attribute_id in attribute_list:
+            if attributes.HoldTime.attribute_id in attribute_list: # check HoldTime conformance

```

Comment:
you are only checking against HoldTimeLimits. Why are we checking for HoldTime attribute existing here? 
Test plan does not seem to enforce that holdtime attribute is within the limits (although maybe it should....)

Reply:
HoldTime is a conformance attribute for HoldTimeLimits.  So, I checked here. I will update the test plan accordingly.

Reply:
The test plan as I have it has the hold time tests against the limits in step 6.
More over , we should test the hold time limits regardless if HoldTime exists or not.

You can have an early exit of `assert_in` to say that we expect HoldTime to exist if limits exist. However as it stands it seems to make no sense to validate the hold time limits only if the attribute hold time exists.

Reply:
I will implement assert_in here as suggested.